### PR TITLE
Fix bug which created dead and incorrectly typed blocks into IR, triggering a verification failure.

### LINKED ISSUE #3205: ICE: Verification failed: Function expiry return type must match its RET instructions
```
error: Internal compiler error: Verification failed: Function expiry return type must match its RET instructions.
Please file an issue on the repository and include the code that triggered this error.
```
Heres the code

```
#[storage(read)]
    fn expiry(name: str[8]) -> Result<u64, RegistrationValidityError> {
        match storage.names.get(name) {
            Option::Some(record) => {
                match timestamp() < record.expiry {
                    true => {
                        return Result::Ok(record.expiry);
                    },
                    false => {
                        return Result::Err(RegistrationValidityError::NameExpired);
                    }
                }
            },
            Option::None => {
                return Result::Err(RegistrationValidityError::NameNotRegistered);
            }
        }
    }
```

with record being of type Record
```
pub struct Record {
    /// The timestamp at which the name expires, and someone else can re-register the same name
    expiry: u64,
    /// The identity to which the name resolves to
    identity: Identity,
    /// The identity which controls the name, and can change the identity and owner
    owner: Identity,
}
```

If a semi colon is added to the end of the match statement, the error changes to 
`^ This path must return a value of type "Result<u64, RegistrationValidityError>" from function "expiry", but it does not.`

### PR DESCRIPTION
**Pull request by otrho**
If both arms of an `if` (or all arms of a `match` which are converted to `if`s) were terminated with `return` then we still created a merge block with an unknown arg, which ended up being dead code and incorrectly typed.

Closes #3205

### CODE REVIEW

**Comment by otrho** [comment]
I'll look at the E2E failures later today.

**Comment by otrho** [comment]
My first attempt was to fix the compilation of `if` exprs but it proved too hard, hence https://github.com/FuelLabs/sway/issues/3211

This attempt just relaxes the IR verifier a bit to ignore the dead code the compiler is producing.  This isn't great -- IMO the compiler should produce well formed code without dead crap and I think we should fix IR gen as per #3211.

**Comment by vaivaswatha** [comment]
#3219 works. Likely it's a temporary fix too and we'll need something like #3211.

I'm going to close that PR and let you / @mohammadfawaz pick what's the right fix here as I'm not very sure myself.

**Review by mohammadfawaz — APPROVED** [review]
Seems reasonable to me! An alternative is to remove dead blocks right after IR gen so we don't have to run a tree search for every verifier call.

File: sway-ir/src/verify.rs [hunk 1]
```diff
@@ -38,9 +38,38 @@ impl Context {
...
-        for block in &function.blocks {
-            self.verify_block(cur_module, function, &self.blocks[block.0])?;
+        // In theory we should just verify every block in the function, and this is what we used to
+        // do.  But there are current limitations in the Sway compiler where it will hackily create
+        // dead code due to poor assumptions about how it should work.
+        //
+        // So for now we'll verify only those blocks which are reachable from `entry`.  This should
+        // be reversed once https://github.com/FuelLabs/sway/issues/3211 is addressed.
+        let mut verified_blocks = rustc_hash::FxHashSet::default();
+        let mut block_queue = vec![function.blocks[0]];
```
[the remark this answers is not in the document]
Comment (otrho) [reply sway-ir/src/verify.rs:48]: Yeah, most of the time the queue wouldn't grow beyond 2 elements, though in my testing I ran this version against `stdlib/vec` and verification was essentially instantaneous, so I don't think we need to specialise.

**Comment by otrho** [comment]
> Seems reasonable to me! An alternative is to remove dead blocks right after (or as the last step of) IR gen so we don't have to run a tree search for every verifier call.

Yep.  Just as hacky though.  None of this is great, I'd really rather see #3211 be the actual fix, and this change can be reverted.

**Comment by vaivaswatha** [comment]
> Seems reasonable to me! An alternative is to remove dead blocks right after (or as the last step of) IR gen so we don't have to run a tree search for every verifier call.

In general, it is *okay* for IR Gen to generate dead blocks or any other inefficient code. It's the optimizers work to optimize that later. What is not okay is for it to generate code that doesn't verify / type check.

**Review by vaivaswatha — APPROVED** [review]

**Comment by otrho** [comment]
OK, I had a chat with @vaivaswatha about `undef` constants and we ended up agreeing that in this case it's OK to use them, and it's probably a better idea than changing the verifier.  #3211 is still needed and maybe the special casing can be reversed then.

**Review by vaivaswatha — APPROVED** [review]

**Review by mohammadfawaz — APPROVED** [review]

**Pull request merged**